Amazon.QLDB.Driver/driver/QldbDriverBase.cs

```diff
@@ -137,12 +136,41 @@ internal void ReleaseSession(T session)
             }
         }
 
-        internal void LogSessionPoolState()
+        internal void ThrowIfClosed()
+        {
+            if (this.IsClosed)
+            {
+                this.Logger.LogError(ExceptionMessages.DriverClosed);
+                throw new QldbDriverException(ExceptionMessages.DriverClosed);
+            }
+        }
+
+        internal T GetSessionFromPool()
         {
             this.Logger.LogDebug(
                 "Getting session. There are {} free sessions and {} available permits.",
                 this.SessionPool.Count,
                 this.SessionPool.BoundedCapacity - this.PoolPermits.CurrentCount);
+
+            if (this.PoolPermits.Wait(0))
+            {
+                lock (this)
+                {
+                    if (this.SessionPool.Count > 0)
```

Comment:
Actually, I think you want `TryTake` instead of Take, then you don't need this lock at all.

Reply:
Good suggestion! Done.